File: FantasyRadio/src/main/java/ru/sigil/fantasyradio/saved/MP3Collection.java
package ru.sigil.fantasyradio.saved;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import javax.inject.Inject;
import javax.inject.Singleton;

import ru.sigil.bassplayerlib.ITrack;
import ru.sigil.bassplayerlib.ITracksCollection;

@Singleton
public class MP3Collection implements ITracksCollection {
    public static final String ARTIST = "ARTIST";
    public static final String TITLE = "TITLE";
    public static final String TIME = "TIME";
    public static final String DIRECTORY = "DIRECTORY";

    private Context context;
    private final Object saveSync = new Object();
    private SQLiteDatabase mDatabase;

    @Inject
    public MP3Collection(Context c) {
        context = c;
        mDatabase = context.openOrCreateDatabase(
                "MP3Base", 0, null);
        mDatabase
                .execSQL("CREATE TABLE IF NOT EXISTS MP3ENTITYES (_id INTEGER PRIMARY KEY AUTOINCREMENT,"
                        + TITLE + " VARCHAR(300),"
                        + ARTIST + " VARCHAR(300),"
                        + TIME + " VARCHAR(300)," + "DIRECTORY VARCHAR(300))");
    }

    public void remove(ITrack mp3entity) {
        synchronized (saveSync) {
            try {
                mDatabase.delete("MP3ENTITYES",
                        DIRECTORY + " = '" + mp3entity.getDirectory() + "'", null);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void add(ITrack mp3entity) {
        synchronized (saveSync) {
            ContentValues cv = new ContentValues();
            cv.put(TITLE, mp3entity.getTitle());
            cv.put(ARTIST, mp3entity.getArtist());
            cv.put(TIME, mp3entity.getTime());
            cv.put(DIRECTORY, mp3entity.getDirectory());
            mDatabase.insert("MP3ENTITYES", null, cv);
        }
    }

    public Cursor getCursor() {
        // Make the query.
        try {
            return mDatabase.query("MP3ENTITYES", null, null,
                    null, null, null, "_id DESC");
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public MP3Entity getNext(ITrack entity) {
        // Make the query.
        if (entity == null)
            return null;
        synchronized (saveSync) {
            try {
                Cursor managedCursor = mDatabase.query("MP3ENTITYES", null, "_id < (SELECT _id FROM MP3ENTITYES WHERE "
                        + DIRECTORY + " = '" + entity.getDirectory() + "')", null, null, null, "_id DESC");
                if (managedCursor != null) {
                    if (managedCursor.moveToFirst()) {
                        MP3Entity mp3Entity = new MP3Entity(managedCursor.getString(
                                managedCursor.getColumnIndex(ARTIST)),
                                managedCursor.getString(managedCursor.getColumnIndex(TITLE)),
                                managedCursor.getString(managedCursor.getColumnIndex(DIRECTORY)),
                                managedCursor.getString(managedCursor.getColumnIndex(TIME)));
                        return mp3Entity;
                    }
                }
                managedCursor.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return null;
    }
}

File: FantasyRadio/src/main/java/ru/sigil/fantasyradio/saved/MP3Collection.kt
package ru.sigil.fantasyradio.saved

import android.content.Context
import ru.sigil.bassplayerlib.ITracksCollection
import android.database.sqlite.SQLiteDatabase
import javax.inject.Singleton
import ru.sigil.bassplayerlib.ITrack
import android.content.ContentValues
import android.database.Cursor
import ru.sigil.fantasyradio.saved.DbConstants.ARTIST
import ru.sigil.fantasyradio.saved.DbConstants.DIRECTORY
import ru.sigil.fantasyradio.saved.DbConstants.TIME
import ru.sigil.fantasyradio.saved.DbConstants.TITLE
import javax.inject.Inject

/**
 * Created by namelessone
 * on 02.12.18.
 */
@Singleton
class MP3Collection @Inject constructor (context: Context): ITracksCollection {
    private var mDatabase: SQLiteDatabase? = null
    private val saveSync = Any()

    init {
        mDatabase = context.openOrCreateDatabase("MP3Base", 0, null)
        mDatabase?.execSQL("CREATE TABLE IF NOT EXISTS MP3ENTITYES (_id INTEGER PRIMARY KEY AUTOINCREMENT,"
                        + TITLE + " VARCHAR(300),"
                        + ARTIST + " VARCHAR(300),"
                        + TIME + " VARCHAR(300)," + "DIRECTORY VARCHAR(300))")
    }

    override fun remove(mp3entity: ITrack) {
        synchronized(saveSync) {
            try {
                mDatabase?.delete("MP3ENTITYES",
                        DIRECTORY + " = '" + mp3entity.directory + "'", null)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }

    override fun add(mp3entity: ITrack) {
        synchronized(saveSync) {
            val cv = ContentValues()
            cv.put(TITLE, mp3entity.title)
            cv.put(ARTIST, mp3entity.artist)
            cv.put(TIME, mp3entity.time)
            cv.put(DIRECTORY, mp3entity.directory)
            mDatabase?.insert("MP3ENTITYES", null, cv)
        }
    }

    fun getCursor(): Cursor? {
        // Make the query.
        try {
            return mDatabase?.query("MP3ENTITYES", null, null, null, null, null, "_id DESC")
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return null
    }

    override fun getNext(entity: ITrack?): MP3Entity? {
        // Make the query.
        if (entity == null)
            return null
        synchronized(saveSync) {
            try {
                val managedCursor = mDatabase?.query("MP3ENTITYES", null, "_id < (SELECT _id FROM MP3ENTITYES WHERE "
                        + DIRECTORY + " = '" + entity.directory + "')", null, null, null, "_id DESC")
                if (managedCursor != null) {
                    if (managedCursor.moveToFirst()) {
                        return MP3Entity(managedCursor.getString(
                                managedCursor.getColumnIndex(ARTIST)),
                                managedCursor.getString(managedCursor.getColumnIndex(TITLE)),
                                managedCursor.getString(managedCursor.getColumnIndex(DIRECTORY)),
                                managedCursor.getString(managedCursor.getColumnIndex(TIME)))
                    }
                }
                managedCursor!!.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return null
    }
}